Improve stability of package and tests

jacekbogdanski opened the pull request:
I've fixed some tests so the whole test suite is more stable now. Also, I've corrected the implementation of `getEditorNamespace` because:

* in case of error, it didn't clean up the promise
* in case of empty string, it directly threw an error, which is not consistent with promise-based functions (you shouldn't have to decorate promised function with try/catch block)

The tests are still failing because of https://github.com/ckeditor/ckeditor4-angular/blob/dc423da5ecc473a834dd6b31fd6b53918f30fe37/src/ckeditor/ckeditor.component.spec.ts#L122-L170 test suite. However, those tests should no longer validate this case, as we already fixed upstream CKEditor 4 issue which required using `divarea` plugin for integrations. I will extract the issue as a separate ticket, as we should clean up the production code also.

## Proposed changelog entry

```
* [#128](https://github.com/ckeditor/ckeditor4-angular/issues/128): Improve the stability of `getEditorNamespace()` method.
```

Closes #121
Closes #129. 
Closes #128.

f1ames commented:
I suppose it closes #93 too?

For anyone merging it, I think we should add changelog entry too, since @jacekbogdanski improved (`getEditorNamespace`) which is a part of he production code. And for the backlog entry we need an issue :joy: :see\_no\_evil: :chains:

jacekbogdanski commented:
> I suppose it closes #93 too?

Yes, of course, looks like I confused ticket numbers.

> For anyone merging it, I think we should add changelog entry too, since @jacekbogdanski improved (getEditorNamespace) which is a part of he production code. And for the backlog entry we need an issue joy see_no_evil chains

I wasn't sure about that because this change doesn't seem to matter from user perspective. But indeed in case if someone changes CDN URL into an empty string she/he will get console warning instead of breaking the whole angular app.

I've extracted ticket for updating failing tests / divarea leftovers: https://github.com/ckeditor/ckeditor4-angular/issues/127

jacekbogdanski commented:
Actually, I checked it right now and I see that a user gets exactly the same error message after proposed changes and the app behaves the same. I'm for skipping changelog entry and separate issue for that.

f1ames commented:
> Actually, I checked it right now and I see that a user gets exactly the same error message after proposed changes and the app behaves the same. I'm for skipping changelog entry and separate issue for that.

Good point, I was thinking more about the improvement itself and that it can solve some real-live edge cases (since now we clear and reject promise correctly, which could cause some issues earlier), so it might be useful info for developers anyway (and additional changelog entry indicates that we are working actively on the integration which is good sign for developers too). WDYT?

Dumluregn commented on src/ckeditor/ckeditor.component.spec.ts:
```diff
@@ -423,7 +416,7 @@ describe( 'CKEditorComponent', () => {
 					} );
 
-					it( 'drag/drop events should emit component dragStart, dragEnd and drop', async done => {
+					it( 'drag/drop events should emit component dragStart, dragEnd and drop', done => {
```
Correct me if I'm wrong, but I suppose this test could also be rewritten to return a promise instead of using `done()` method? Then all tests would be nicely unified.

jacekbogdanski replied:
It seems like this function shouldn't be async at all. The whole code is synchronous and we are not waiting for editor data.

Dumluregn commented on src/ckeditor/ckeditor.component.spec.ts:
```diff
@@ -4,12 +4,11 @@
 import {
 	fireDragEvent,
 	mockDropEvent,
-	mockPasteEvent,
```
This helper can be removed from `test.tools.js` file as it isn't used anywhere else.

Dumluregn commented on src/ckeditor/ckeditor.helpers.spec.ts:
```diff
@@ -52,6 +53,11 @@ describe( 'getEditorNamespace', () => {
 
 	it( 'returns the same promise', () => {
-		expect( getEditorNamespace( fakeScript ) ).toBe( getEditorNamespace( fakeScript ) );
+		const promise1 =  getEditorNamespace( fakeScript );
+		const promise2 =  getEditorNamespace( fakeScript );
```
```suggestion
		const promise1 = getEditorNamespace( fakeScript );
		const promise2 = getEditorNamespace( fakeScript );
```

Dumluregn commented on src/ckeditor/ckeditor.component.spec.ts:
```diff
@@ -4,12 +4,11 @@
  */
 
 import waitUntil from 'wait-until-promise';
-import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
I see you removed `async` module, but some tests are still async - do I understand correctly that this was never necessary here?

jacekbogdanski replied:
Yes, because this [code is passing promises](https://angular.io/api/core/testing/TestBed#compilecomponents), so we can use that instead.

Dumluregn commented:
Created #128 to provide a changelog entry here.

jacekbogdanski commented:
> PR lacks the change of karma.conf.js removing the unrandomising snippet:

I'm not sure about that. Test still sometimes fails, probably also because of https://github.com/ckeditor/ckeditor4-angular/issues/127

I propose to do that once we fix #127 and we are sure that CI will be green with random order.

jacekbogdanski commented:
> Well, the console.warn tests are failing because you fixed them smile So CI will be red for now no matter if you run tests in order or randomly. So are we good with adding this Karma config? Or you meant something different?

Yeah, they are failing despite random order, so let's make it random by default.

Dumluregn commented:
LGTM!

jacekbogdanski commented:
We come to the conclusion that it takes too much time to fix running tests in random order. Instead, we will just focus on improving overall asynchronous test correctness proposed in this PR. I've changed `Closes`, so it will close correct issues. I've also reverted changes to karma config.
